File: androidApp/v2/feature/search/src/main/java/com/deadly/v2/feature/search/screens/main/SearchShortcuts.kt
package com.deadly.v2.feature.search.screens.main

import androidx.annotation.DrawableRes

data class SearchShortcut(
    val title: String,
    val subtitle: String,
    val searchQuery: String,
    val priority: Int = 0,
    @DrawableRes val discoverImageRes: Int? = null,
    @DrawableRes val browseImageRes: Int? = null,
)

val allSearchShortcuts = listOf(
    // Filters (priority 10) — these use the new FTS tags
    SearchShortcut("Top Rated", "Highest rated recordings", "top-rated", 10),
    SearchShortcut("Popular", "Most reviewed shows", "popular", 10),
    SearchShortcut("Soundboard", "Direct from the mixing board", "sbd", 10),
    SearchShortcut("Audience", "Taped from the crowd", "aud", 10),

    // Venues (priority 5)
    SearchShortcut("Fillmore", "Fillmore East & West", "Fillmore", 5),
    SearchShortcut("Winterland", "San Francisco's legendary venue", "Winterland", 5),
    SearchShortcut("Red Rocks", "Morrison, Colorado", "Red Rocks", 5),
    SearchShortcut("MSG", "Madison Square Garden", "Madison Square Garden", 5),
    SearchShortcut("Capitol Theatre", "Port Chester, New York", "Capitol Theatre", 5),
    SearchShortcut("Barton Hall", "Cornell University, Ithaca", "Barton Hall", 5),

    // Cities (priority 5)
    SearchShortcut("New York", "The Big Apple", "New York", 5),
    SearchShortcut("San Francisco", "Home turf", "San Francisco", 5),
    SearchShortcut("Chicago", "The Windy City", "Chicago", 5),
    SearchShortcut("Philadelphia", "City of Brotherly Love", "Philadelphia", 5),
    SearchShortcut("Boston", "New England shows", "Boston", 5),
    SearchShortcut("Los Angeles", "SoCal shows", "Los Angeles", 5),

    // Songs (priority 5)
    SearchShortcut("Dark Star", "The quintessential jam", "Dark Star", 5),
    SearchShortcut("Scarlet > Fire", "The classic combo", "Scarlet Begonias", 5),
    SearchShortcut("Wharf Rat", "The ballad of August West", "Wharf Rat", 5),
    SearchShortcut("Truckin'", "What a long strange trip", "Truckin", 5),
    SearchShortcut("Eyes of the World", "Jazz-infused Garcia classic", "Eyes of the World", 5),
    SearchShortcut("Sugar Magnolia", "Sunshine daydream", "Sugar Magnolia", 5),

    // Members (priority 3)
    SearchShortcut("Brent Era", "Brent Mydland on keys", "Brent", 3),
    SearchShortcut("Pigpen Era", "Blues and soul frontman years", "Pigpen", 3),
    SearchShortcut("Keith Era", "Keith Godchaux on piano", "Keith", 3),
)

File: iosApp/deadly/Feature/Search/SearchShortcuts.swift
import SwiftUI

struct SearchShortcut: Identifiable {
    let id = UUID()
    let title: String
    let subtitle: String
    let searchQuery: String
    let priority: Int
    let category: ShortcutCategory
}

enum ShortcutCategory: String, CaseIterable {
    case filter, venue, city, song, member

    var displayName: String {
        switch self {
        case .filter: return "Filter"
        case .venue: return "Venue"
        case .city: return "City"
        case .song: return "Song"
        case .member: return "Member"
        }
    }

    var systemImage: String {
        switch self {
        case .filter: return "line.3.horizontal.decrease.circle"
        case .venue: return "building.2"
        case .city: return "mappin.and.ellipse"
        case .song: return "music.note"
        case .member: return "person"
        }
    }

    var tintColor: Color {
        switch self {
        case .filter: return DeadlyColors.primary
        case .venue: return DeadlyColors.secondary
        case .city: return DeadlyColors.tertiary
        case .song: return .purple
        case .member: return .orange
        }
    }
}

let allSearchShortcuts: [SearchShortcut] = [
    // Filters (priority 10) - these use FTS tags
    SearchShortcut(title: "Top Rated", subtitle: "Highest rated recordings", searchQuery: "top-rated", priority: 10, category: .filter),
    SearchShortcut(title: "Popular", subtitle: "Most reviewed shows", searchQuery: "popular", priority: 10, category: .filter),
    SearchShortcut(title: "Soundboard", subtitle: "Direct from the board", searchQuery: "sbd", priority: 10, category: .filter),
    SearchShortcut(title: "Audience", subtitle: "Taped from the crowd", searchQuery: "aud", priority: 10, category: .filter),

    // Venues (priority 5)
    SearchShortcut(title: "Fillmore", subtitle: "Fillmore East & West", searchQuery: "Fillmore", priority: 5, category: .venue),
    SearchShortcut(title: "Winterland", subtitle: "San Francisco", searchQuery: "Winterland", priority: 5, category: .venue),
    SearchShortcut(title: "Red Rocks", subtitle: "Morrison, CO", searchQuery: "Red Rocks", priority: 5, category: .venue),
    SearchShortcut(title: "MSG", subtitle: "Madison Square Garden", searchQuery: "Madison Square Garden", priority: 5, category: .venue),
    SearchShortcut(title: "Capitol Theater", subtitle: "Port Chester, NY", searchQuery: "Capitol Theater", priority: 5, category: .venue),
    SearchShortcut(title: "Barton Hall", subtitle: "Cornell University", searchQuery: "Barton Hall", priority: 5, category: .venue),

    // Cities (priority 5)
    SearchShortcut(title: "New York", subtitle: "The Big Apple", searchQuery: "New York", priority: 5, category: .city),
    SearchShortcut(title: "San Francisco", subtitle: "Home turf", searchQuery: "San Francisco", priority: 5, category: .city),
    SearchShortcut(title: "Chicago", subtitle: "The Windy City", searchQuery: "Chicago", priority: 5, category: .city),
    SearchShortcut(title: "Philadelphia", subtitle: "City of Brotherly Love", searchQuery: "Philadelphia", priority: 5, category: .city),
    SearchShortcut(title: "Boston", subtitle: "New England shows", searchQuery: "Boston", priority: 5, category: .city),
    SearchShortcut(title: "Los Angeles", subtitle: "SoCal shows", searchQuery: "Los Angeles", priority: 5, category: .city),

    // Songs (priority 5)
    SearchShortcut(title: "Dark Star", subtitle: "The quintessential jam", searchQuery: "Dark Star", priority: 5, category: .song),
    SearchShortcut(title: "Scarlet > Fire", subtitle: "The classic combo", searchQuery: "Scarlet Begonias", priority: 5, category: .song),
    SearchShortcut(title: "Wharf Rat", subtitle: "Ballad of August West", searchQuery: "Wharf Rat", priority: 5, category: .song),
    SearchShortcut(title: "Truckin'", subtitle: "What a long strange trip", searchQuery: "Truckin", priority: 5, category: .song),
    SearchShortcut(title: "Eyes of the World", subtitle: "Jazz-infused Garcia", searchQuery: "Eyes of the World", priority: 5, category: .song),
    SearchShortcut(title: "Sugar Magnolia", subtitle: "Sunshine daydream", searchQuery: "Sugar Magnolia", priority: 5, category: .song),

    // Members (priority 3)
    SearchShortcut(title: "Brent Era", subtitle: "Brent Mydland on keys", searchQuery: "Brent", priority: 3, category: .member),
    SearchShortcut(title: "Pigpen Era", subtitle: "Blues and soul years", searchQuery: "Pigpen", priority: 3, category: .member),
    SearchShortcut(title: "Keith Era", subtitle: "Keith Godchaux on piano", searchQuery: "Keith", priority: 3, category: .member),
]

/// 3 discover cards, rotated every 4 hours via seeded shuffle.
/// Pass refreshCounter to force re-roll on pull-to-refresh.
func discoverShortcuts(refreshCounter: Int = 0) -> [SearchShortcut] {
    rotatedSubset(of: allSearchShortcuts, count: 3, seedOffset: refreshCounter * 1000)
}

/// 8 browse cards with priority >= 5, rotated every 4 hours via seeded shuffle.
/// Pass refreshCounter to force re-roll on pull-to-refresh.
/// Excludes items already shown in discoverShortcuts to avoid duplicates.
func browseShortcuts(refreshCounter: Int = 0) -> [SearchShortcut] {
    let discoverItems = discoverShortcuts(refreshCounter: refreshCounter)
    let discoverTitles = Set(discoverItems.map(\.title))
    let eligible = allSearchShortcuts.filter {
        $0.priority >= 5 && !discoverTitles.contains($0.title)
    }
    return rotatedSubset(of: eligible, count: 8, seedOffset: 500 + refreshCounter * 1000)
}

private func rotatedSubset(of shortcuts: [SearchShortcut], count: Int, seedOffset: Int = 0) -> [SearchShortcut] {
    let seed = Int(Date().timeIntervalSince1970 / (4 * 3600)) + seedOffset
    var rng = SeededRandomNumberGenerator(seed: UInt64(seed))
    var shuffled = shortcuts
    shuffled.shuffle(using: &rng)
    return Array(shuffled.prefix(count))
}

private struct SeededRandomNumberGenerator: RandomNumberGenerator {
    private var state: UInt64

    init(seed: UInt64) {
        // Mix the seed using splitmix64 to ensure good bit distribution
        var z = seed &+ 0x9e3779b97f4a7c15
        z = (z ^ (z >> 30)) &* 0xbf58476d1ce4e5b9
        z = (z ^ (z >> 27)) &* 0x94d049bb133111eb
        state = z ^ (z >> 31)
        // Ensure state is never 0 (xorshift produces only zeros from zero state)
        if state == 0 { state = 1 }
    }

    mutating func next() -> UInt64 {
        // xorshift64
        state ^= state << 13
        state ^= state >> 7
        state ^= state << 17
        return state
    }
}
